src/Symfony/Component/Security/Http/Authenticator/HttpBasicAuthenticator.php

```diff
@@ -0,0 +1,100 @@
+ */
+
+namespace Symfony\Component\Security\Http\Authenticator;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\AuthenticationServiceException;
+use Symfony\Component\Security\Core\User\PasswordUpgraderInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\PasswordUpgradeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\PasswordCredentials;
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Component\Security\Http\EntryPoint\AuthenticationEntryPointInterface;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class HttpBasicAuthenticator implements AuthenticatorInterface, AuthenticationEntryPointInterface
+{
+    private $realmName;
+    private $userProvider;
+    private $logger;
+
+    public function __construct(string $realmName, UserProviderInterface $userProvider, ?LoggerInterface $logger = null)
+    {
+        $this->realmName = $realmName;
+        $this->userProvider = $userProvider;
+        $this->logger = $logger;
+    }
+
+    public function start(Request $request, AuthenticationException $authException = null)
+    {
+        $response = new Response();
+        $response->headers->set('WWW-Authenticate', sprintf('Basic realm="%s"', $this->realmName));
+        $response->setStatusCode(401);
+
+        return $response;
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return $request->headers->has('PHP_AUTH_USER');
+    }
+
+    public function authenticate(Request $request): PassportInterface
+    {
+        $username = $request->headers->get('PHP_AUTH_USER');
+        $password = $request->headers->get('PHP_AUTH_PW', '');
+
+        $user = $this->userProvider->loadUserByUsername($username);
+        if (!$user instanceof UserInterface) {
+            throw new AuthenticationServiceException('The user provider must return a UserInterface object.');
+        }
+
+        $badges = [];
+        if ($this->userProvider instanceof PasswordUpgraderInterface) {
+            $badges[] = new PasswordUpgradeBadge($password, $this->userProvider);
+        }
+
+        return new Passport($user, new PasswordCredentials($password), $badges);
+    }
+
+    public function createAuthenticatedToken(PassportInterface $passport, $providerKey): TokenInterface
```

Comment:
Same here: some phpdoc (at least) to say this will be a `Passport` object with a `getUser()` method.